### PR TITLE
Refactor event handling to fix stuck application issues

### PR DESCRIPTION
**Pull request by NohaIhab**
## Summary:
- simplified event handlers by removing update_container methods and deferred error handling to main method
- called main (on_event) method at the end of *_pebble_ready events
- replaced CheckFailed errors with ErrorWithStatus to be caught in main method

### CODE REVIEW

**Comment by i-chvets** [comment]
@NohaIhab @ca-scribner Just wanted to add a comment to record my confusion.
In multi-container scenario we test that install event can come after pebble ready event. Are we sure that is what is happening in real deployment?
In the test we are setting leader to True, before install event, but that is not what should be happening in real deployment. According to [charm's life](https://juju.is/docs/sdk/a-charms-life) leadership selection happens after install.
As a result, pebble ready event should always come after install: install->leader-set->pebble-ready. 
However, in the same doc it says that pebble ready event can happen at any time of charm's life cycle.
Another [pebble ready specific doc](https://juju.is/docs/sdk/container-name-pebble-ready-event) also supports install->pebble-ready relationship.
The fix and test LGTM.
The testing for different even sequences is great and we should adapt this to other charms.

**Comment by NohaIhab** [comment]
@i-chvets 

- You're absolutely correct, the issue was due to pebble_ready being called with no other hooks after so just checking that after it's called the charm reaches Active is enough. Does that make sense?
- I agree on the set_leader placement compared to real deployment, however it's in the ops documentation that set_leader is called before begin_with_initial_hooks which triggers install [here](https://ops.readthedocs.io/en/latest/#ops.testing.Harness.begin_with_initial_hooks), not sure why.

File: src/charm.py [hunk 1]
```diff
@@ -63,21 +63,21 @@ def __init__(self, *args):
...
-            self.on.leader_elected,
-            self.on.upgrade_charm,
-            self.on["kubeflow-profiles"].relation_changed,
-            self.on.config_changed,
-        ]:
-            self.framework.observe(event, self.main)
+        # setup events to be handled by specific event handlers
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.remove, self._on_remove)
+        self.framework.observe(self.on.config_changed, self.service_patcher._patch)
```
Comment (i-chvets) [inline src/charm.py:76]: At line 68 above `config_changed` is setup with `_on_event`. We need to review this.

Comment (NohaIhab) [reply src/charm.py:76]: I think an event can have multiple observers, how can we check on that?

Comment (i-chvets) [reply src/charm.py:76]: Yes, it looks like we can have multiple observers.

Comment (i-chvets) [reply src/charm.py:76]: Are we happy with the order of those observers? I.e. order they will be called when event is triggered.

Comment (NohaIhab) [reply src/charm.py:76]: patch first then main would make sense. will adjust

**Comment by ca-scribner** [comment]
@i-chvets you raise a good point about the leadership event.  I *think* you're correct that `install` will always precede `leader-elected`, and that those hooks may happen before or after any `pebble-ready` events.  Which I think means there's yet another sequence we mishandled and been getting lucky on :) .  

For example, what about:
* all the `pebble-ready` events fire and set up their containers, but even if we then call `main` 
* `install` fires.  If it tried to run `main` for the charm to go active, it would fail because it is not the leader
* ?

What I think saves us in this case is that, after `install` fires, a `config-changed` event is queued but it (always? sometimes?) ends up queued behind the leader-elected event.  See this sequence from deploying the charm yesterday:
```
Time                        Type       Status       Message
02 Feb 2023 12:48:12-05:00  juju-unit  allocating   
02 Feb 2023 12:48:12-05:00  workload   waiting      installing agent
02 Feb 2023 12:48:22-05:00  workload   waiting      agent initializing
02 Feb 2023 12:48:28-05:00  juju-unit  executing    running install hook  <------------if we tried to go active, we'd fail leadership check
02 Feb 2023 12:48:30-05:00  juju-unit  executing    running kubeflow-profiles-relation-created hook<------------if we tried to go active, we'd fail leadership check
02 Feb 2023 12:48:36-05:00  juju-unit  executing    running kubeflow-profiles-relation-created hook<------------if we tried to go active, we'd fail leadership check
02 Feb 2023 12:48:37-05:00  juju-unit  executing    running leader-elected hook
02 Feb 2023 12:48:39-05:00  juju-unit  executing    running kubeflow-kfam-pebble-ready hook <---we are leader, but (previously we never tried to go active)
02 Feb 2023 12:48:40-05:00  juju-unit  executing    running config-changed hook <--- this event is what actually puts us into active, because we also subscribe it to `main`
02 Feb 2023 12:48:42-05:00  juju-unit  executing    running start hook
```

Our charm is actually going active from the `config-changed`.  And if by chance we had the order:
* `pebble(s)-ready`
* `install`
* `config-changed`  <--- not yet leader, so charm wont go active
* `leader-elected`

then I think we'd still be stuck with charm not in active state.  After reading the [leader-elected](https://juju.is/docs/sdk/leader-elected-event) and [config-changed](https://juju.is/docs/sdk/config-changed-event) docs, I'm not sure if this is a valid sequence of events or not but maybe we should be defensive against it just in case?

File: src/charm.py (hunk 1 above)
Comment (i-chvets) [reply src/charm.py:69]: I suspect `_patch` is added to observer's list automatically when you instantiate `KubernetesServicePatch`:
https://github.com/canonical/kubeflow-profiles-operator/blob/main/lib/charms/observability_libs/v1/kubernetes_service_patch.py#L211

I think, we do not need to explicitly add.

**Comment by i-chvets** [comment]
When charm is first deployed the sequence you showed should not happen according to [this doc](https://juju.is/docs/sdk/container-name-pebble-ready-event).
Quoting:
***
Emission sequence
When **a charm is first deployed** or scaled up, **`pebble-ready` is guaranteed to fire only after `start`** (consequentially, it fires only after install and config-changed fired). If relation-created and leader-elected/leader-settings-changed were ready to fire before start, then pebble-ready is also guaranteed to fire only after relation-created and leadership hooks have fired.
SCENARIO	| EXAMPLE COMMAND|	RESULTING EVENTS
--- | --- | ---
Create unit |	juju deploy foo | `install -> config-changed -> start -> bar-pebble-ready`
***
I guess, after it deployed `pebble-ready` can happen any time.

**Comment by i-chvets** [comment]
We currently ignore `leader-elected` event. May be we should add an observer to it to make sure we switch to Active when we are a leader and container is available?

**Review by i-chvets — APPROVED** [review]
Approved.
Will address anomalies in event sequence in different item.

**Pull request merged**